src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+                absl::Nanoseconds(client_config.presumed_abort_time().nanos())
+          : current_time + default_transaction_duration;
+  if (presumed_abort_time < current_time) {
+    return absl::InvalidArgumentError("Presumed abort time is in the past");
+  }
+  common::TransactionConfig final_config;
+  timespec presumed_abort_timespec = absl::ToTimespec(presumed_abort_time);
+  final_config.mutable_presumed_abort_time()->set_seconds(
+      presumed_abort_timespec.tv_sec);
+  final_config.mutable_presumed_abort_time()->set_nanos(
+      presumed_abort_timespec.tv_nsec);
+  return final_config;
+}
+
+absl::StatusOr<const std::string> GetTransactionId(
+    const std::string &client_transaction_id,
+    const std::string &client_address) {
+  std::array<unsigned char, 32> hash_value;
+  if (!HashSha256(absl::StrCat(client_transaction_id, "~", client_address),
+                  hash_value.data())) {
+    return absl::UnknownError(
+        "Could not hash client transaction id and address");
+  }
+  std::stringstream ss;
+  for (const auto &hash_byte : hash_value) {
+    ss << absl::StrCat(absl::Hex(hash_byte, absl::kZeroPad2));
+  }
+  return ss.str();
+}
+
+SplitTransaction SplitClientTransaction(
+    const common::Transaction &original_transaction) {
+  absl::flat_hash_map<std::string, TransactionWithNamespace>
+      transactions_by_namespace;
+  bool has_writes = false;
+  bool has_reads = false;
+  for (const auto &operation : original_transaction.ops()) {
+    TransactionWithNamespace &transaction_with_namespace =
+        transactions_by_namespace[operation.namespace_().address()];
+    transaction_with_namespace.ns = operation.namespace_();
+    *transaction_with_namespace.transaction.add_ops() = operation;
+    if (operation.has_put()) {
+      has_writes = true;
+    } else {
+      has_reads = true;
+    }
+    switch (transaction_with_namespace.transaction_type) {
+      case UNSET_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? WRITE_ONLY_TRANSACTION
+                                : READ_ONLY_TRANSACTION;
+        break;
+      case READ_ONLY_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? READ_WRITE_TRANSACTION
+                                : READ_ONLY_TRANSACTION;
+        break;
+      case WRITE_ONLY_TRANSACTION:
+        transaction_with_namespace.transaction_type =
+            operation.has_put() ? WRITE_ONLY_TRANSACTION
+                                : READ_WRITE_TRANSACTION;
+        break;
+      case READ_WRITE_TRANSACTION:
+        break;
+    }
+  }
+  std::vector<TransactionWithNamespace> split_transactions;
+  split_transactions.reserve(transactions_by_namespace.size());
+  for (const auto &transaction_with_namespace : transactions_by_namespace) {
+    split_transactions.push_back(transaction_with_namespace.second);
+  }
+  std::sort(split_transactions.begin(), split_transactions.end(),
```

Comment:
Can you add some comment here on why sorting is needed? I think we are trying to avoid cycles in lock dependency graph which may lead to transaction timeouts?

Reply:
That's a good point. I added it so that it's deterministic which cohort is which index for the blockchain code, but I guess there's no reason it has to be deterministic. I'll remove the sorting.

Reply:
I changed it to be test only to make testing deterministic, but it doesn't run in the main version.